Backend guide

jcrist1 opened the pull request:
I feel like I'm getting a little bit annoying now, constantly pushing stuff 😅  but I wanted to get my own thoughts and experiences down building the kotlin backend, before I forget. I created a document which outlines how I would have done it starting again. There are definitely some things that I would have done differently, and I wanted to jot them down. In particular the `insta` tests, but also a more structured approach to initial development. @Manishearth is something like this  useful? If it's not really what you imagine for the backend guide, I'm happy to close, change, or move somewhere else.

jcrist1 committed:
backend-guide - simple set up

jcrist1 committed:
backend-guide

jcrist1 committed:
backend-guide - clean up text

Manishearth commented:
Oh, thank you very much! I'll probably have some comments but it's great to have something like this, this book needs a lot of work especially in the new HIR world with all the new types.

jcrist1 commented:
😄 Cool! I got all of my thoughts out already and looks like I've finally configured my GH notifications to see comments on PRs I'm participating in. So if/when you've got some comments I can address them.

Manishearth commented on src/SUMMARY.md:
```diff
@@ -11,4 +11,5 @@
   - [Lifetimes](lifetimes.md)
-- [Backend developer guide](developer.md)
\ No newline at end of file
+- [Backend developer guide](developer.md)
```
thought: should this file just be the toplevel guide?

Manishearth commented on src/backend_dev_setup.md:
```diff
@@ -0,0 +1,215 @@
+
+Your backend should iterate over all [TypeDefs](https://docs.rs/diplomat_core/latest/diplomat_core/hir/enum.TypeDef.html)
+and generate the required code for these. A good starting point is to create a test for generating a simple opaque struct
+without any methods. Your backend should go in the tool create: create a module `tool/src/{backend}/mod.rs` 
```
```suggestion
without any methods. Your backend should go in the tool crate: create a module `tool/src/{backend}/mod.rs` 
```

Manishearth commented on src/backend_dev_setup.md:
```diff
@@ -0,0 +1,215 @@
+which you can use to pick up your generated code.
+
+## How to Generate the Dylib
+Now to actually test native methods you will need to create a dynamically linked library 
```
issue: not always, no. say that you will need to compile _some_ kind of library: a staticlib, dylib, or wasm lib. And then say here we'll make a dylib.

Manishearth commented on src/backend_dev_setup.md:
```diff
@@ -0,0 +1,215 @@
+  - [ ] owned slices should be kotlin arrays
+  - [ ] slices of strings
+  - [ ] strings
+- [ ] borrows. This is probably one of the trickiest things., as you need to ma
```
issue: this is cut off

Manishearth reviewed (commented):
seems okay, but could have more. still much better than what we had!

jcrist1 committed:
Update src/backend_dev_setup.md

Co-authored-by: Manish Goregaokar <manishsmail@gmail.com>

jcrist1 committed:
backend-guide - address review comments

jcrist1 committed:
Merge branch 'backend-guide' of github.com:jcrist1/book into backend-guide

jcrist1 committed:
backend-guide - more in-depth

jcrist1 committed:
backend-guide

jcrist1 committed:
backend-guide - rename file

jcrist1 committed:
backend-guide - link to method

jcrist1 committed:
backend-guide - consistent type link formatting

jcrist1 committed:
backend - commas

jcrist1 commented:
Thanks for the comments. I replaced the existing backend guide with this, but paraphrased some of the explanation that was there into this document, commented that there are many library options as a compilation target, and fixed the broken comment about borrows.

Manishearth commented on src/developr.md:
```diff
@@ -0,0 +1,235 @@
+# How to Set up a Minimal Backend
```
nit: the file name is wrong

jcrist1 replied:
fixed. Thanks for the review

jcrist1 committed:
backend-guide
